### DISCUSSION TITLE
Controlling LEDs with CPP

### DISCUSSION THREAD

**Discussion by mhl787156** [bitcraze/discussions#1708] [Q&A]
Hi! 

I'm looking at Crazyflie.h (It is avaliable in GitHub: https://github.com/whoenig/crazyflie_cpp/tree/master) and trying to work out how I might be able to control the LEDs but I am rather confused. 

Does anybody have a guide to this CPP library, or some example cpp code for the LED deck? 

I can't make heads or tails comparing to the cfclient python api

Looking to use aerostack2 with LEDs, but will need to extend their crazyflie interface.

**Comment by ToveRumar** [comment]
Hi @mhl787156 !

Do you specifically need to control the LEDs using the Crazyflie FW from whoenig? Or could you use the official C firmware ( https://github.com/bitcraze/crazyflie-firmware)? 
 I would suggest to have a look at how we control the LED ring deck in the official FW; https://github.com/bitcraze/crazyflie-firmware/blob/master/src/deck/drivers/src/ledring12.c

You could just use the same driver to control the LEDs or translate what you need to CPP. 

If this question is more aimed at the https://github.com/whoenig/crazyflie_cpp/ repo I think you might be more in luck by asking in that forum instead. 

I hope I understood your question correctly.

**Comment by mhl787156** [comment]
Hi @ToveRumar ! 

Thanks for your reply! The use case is offboard control of the LED Rings, and extending the library I'm using (aerostack2) to support the LED Rings. They have a cpp interface which uses Whoenig's crazieflie_cpp as the interfacing adapter. I'm assuming that crazyflie_cpp sends commands via the radio which then call functionality from within the onboard crazyflie firmware you linked? 

Since that library is relatively old right now, I wasn't sure if it was still being supported, hence asking here! 

The other reason I was asking here was that in the cfclient python api examples, the memory buffer representing the LED Rings is requested, and then modifying that buffer magically changes the colours of the LED. crazyflie_cpp seems like it might do something similar judging from the variables. So I was hoping to find out a bit more about how the LED Rings are represented within your python library to see if I can apply the same logic to demystify crazyflie_cpp! 

Or I might just have misunderstood the crazyflie-firmware and how I can interact with the LEDs!

**Reply by ToveRumar** [reply]
Hi!

Ok,  now I understand what you mean. 

Yes, this is how the lib works to control the LEDs!  

I would say going about it as you say, by trying to compare it to our lib is your best bet. 
Im afraid I cant give you much more guidance than this.  Lets hope someone in the community has done something similar.
